src/main/java/de/tum/in/www1/artemis/service/TutorialGroupService.java

```diff
@@ -85,6 +98,188 @@ public Set<StudentDTO> registerMultipleStudents(TutorialGroup tutorialGroup, Set
+
+        // === Step 2: If the registration contains a student, try to find a user in the database with the mentioned registration number ===
+        var registrationWithUserIdentifier = registrationsWithTitle.stream().filter(registration -> {
+            if (registration.student() == null) {
+                return false;
+            }
+            boolean hasRegistrationNumber = registration.student().getRegistrationNumber() != null && !registration.student().getRegistrationNumber().isBlank();
+            boolean hasLogin = registration.student().getLogin() != null && !registration.student().getLogin().isBlank();
+            return hasRegistrationNumber || hasLogin;
+        }).collect(Collectors.toSet());
+
+        var registrationsWithMatchingUsers = filterOutWithoutMatchingUser(course, registrationWithUserIdentifier, failedRegistrations);
+        var uniqueRegistrationsWithMatchingUsers = filterOutMultipleRegistrationsForSameUser(registrationsWithMatchingUsers, failedRegistrations);
+
+        // === Step 3: Register all found users to their respective tutorial groups ===
+        var tutorialGroupToRegisteredUsers = new HashMap<TutorialGroup, Set<User>>();
+        for (var registrationUserPair : uniqueRegistrationsWithMatchingUsers) {
+            var tutorialGroup = tutorialGroupTitleToTutorialGroup.get(registrationUserPair.getFirst().title().trim()).get(0);
+            var user = registrationUserPair.getSecond();
+            tutorialGroupToRegisteredUsers.computeIfAbsent(tutorialGroup, key -> new HashSet<>()).add(user);
+        }
+
+        // deregister all students that should be registered to a new tutorial group from their old tutorial groups
+        deregisterMultipleStudentsFromAllTutorialGroupInCourse(uniqueRegistrationsWithMatchingUsers.stream().map(Pair::getSecond).collect(Collectors.toSet()), course,
+                TutorialGroupRegistrationType.INSTRUCTOR_REGISTRATION);
+
+        for (var tutorialGroupAndRegisteredUsers : tutorialGroupToRegisteredUsers.entrySet()) {
+            registerMultipleStudentsToTutorialGroup(tutorialGroupAndRegisteredUsers.getValue(), tutorialGroupAndRegisteredUsers.getKey(),
+                    TutorialGroupRegistrationType.INSTRUCTOR_REGISTRATION);
+        }
+
+        // === Step 4: Create the result for the successful and failed imports ===
+        var result = new HashSet<TutorialGroupRegistrationImportDTO>();
+
+        var successfulImports = new HashSet<>(registrations);
+        successfulImports.removeAll(failedRegistrations);
+        for (var successfulImport : successfulImports) {
+            result.add(successfulImport.withImportResult(true, null));
+        }
+        result.addAll(failedRegistrations);
+
+        return result;
+    }
+
+    private Set<Pair<TutorialGroupRegistrationImportDTO, User>> filterOutMultipleRegistrationsForSameUser(
```

Comment:
Idea: A HashSet of Pairs is more or less a HashTable with these two as key/value. Maybe this makes the code a bit cleaner

Reply:
Good idea! I changed it to a map based implementation in a016d181c6d43157329d66eac85a32c9932e89df